worstcase/state.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// Package worstcase provides the worst-case balance and nonce tracking needed
+// to safely include transactions that are guaranteed to be valid during
+// execution.
+package worstcase
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/state"
+	"github.com/ava-labs/libevm/core/txpool"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/params"
+	"github.com/holiman/uint256"
+
+	"github.com/ava-labs/strevm/gastime"
+	"github.com/ava-labs/strevm/hook"
+	saeparams "github.com/ava-labs/strevm/params"
+)
+
+// State tracks the worst-case gas price and account state as operations are
+// executed.
+//
+// Usage of the [State] should follow the pattern:
+//  1. [State.StartBlock] for each block to be included.
+//  2. [State.GasLimit] and [State.BaseFee] to query the block's parameters.
+//  3. [State.ApplyTx] or [State.Apply] for each transaction or operation to
+//     include in the block.
+//  4. [State.FinishBlock] to finalize the block's gas time.
+//  5. Repeat from step 1 for the next block.
+type State struct {
+	hooks  hook.Points
+	config *params.ChainConfig
+
+	db    *state.StateDB
+	clock *gastime.Time
+
+	qSize, blockSize, maxBlockSize gas.Gas
+
+	baseFee *uint256.Int
+	curr    *types.Header
+	rules   params.Rules
+	signer  types.Signer
+}
+
+// NewState constructs a new worst-case state.
+//
+// The [state.StateDB] MUST be opened at the state immediately following the
+// last-executed block upon which the worst-case state is built. Similarly, the
+// [gastime.Time] MUST be a clone of the gas clock at the same point. The
+// StateDB will only be used as a scratchpad for tracking accounts, and will NOT
+// be committed.
+func NewState(
+	hooks hook.Points,
+	config *params.ChainConfig,
+	db *state.StateDB,
+	fromExecTime *gastime.Time,
+) *State {
+	return &State{
+		hooks:  hooks,
+		config: config,
+		db:     db,
+		clock:  fromExecTime,
+	}
+}
+
+const rateToMaxBlockSize = saeparams.Tau * saeparams.Lambda
+
+var (
+	errNonConsecutiveBlocks = errors.New("non-consecutive block numbers")
+	errQueueFull            = errors.New("queue full")
+)
+
+// StartBlock updates the worst-case state to the beginning of the provided
+// block.
+//
+// It is not necessary for [types.Header.GasLimit] or [types.Header.BaseFee] to
+// be set.
+//
+// If the queue is too full to accept another block, an error is returned.
+func (s *State) StartBlock(hdr *types.Header) error {
+	if c := s.curr; c != nil {
```

Comment:
This check will be unnecessary (and hence the internal one will be stronger) if we accept the hypothetically last-settled block in `NewState()`.